File: app/src/main/java/com/wafflestudio/snutt_staging/ui/TableCreateActivity.java
package com.wafflestudio.snutt_staging.ui;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.NumberPicker;
import android.widget.Toast;

import com.wafflestudio.snutt_staging.R;
import com.wafflestudio.snutt_staging.SNUTTBaseActivity;
import com.wafflestudio.snutt_staging.manager.TableManager;
import com.wafflestudio.snutt_staging.model.Table;

import java.util.List;

import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

/**
 * Created by makesource on 2016. 3. 1..
 */
public class TableCreateActivity extends SNUTTBaseActivity {

    private int year = 2014;
    private int semester = 1;
    private EditText titleText;
    private NumberPicker yearPicker;
    private NumberPicker semesterPicker;
    private Button submitButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityList.add(this);
        setContentView(R.layout.activity_table_create);

        yearPicker = (NumberPicker) findViewById(R.id.yearPicker);
        semesterPicker = (NumberPicker) findViewById(R.id.semesterPicker);
        titleText = (EditText) findViewById(R.id.tableName);

        String[] years = { "2014 년", "2015 년", "2016 년" };
        yearPicker.setMinValue(2014);
        yearPicker.setMaxValue(2016);
        yearPicker.setDisplayedValues(years);
        yearPicker.setOnValueChangedListener(new NumberPicker.OnValueChangeListener() {
            @Override
            public void onValueChange(NumberPicker picker, int oldVal, int newVal) {
                year = newVal;
            }
        });

        final String[] semesters = { "1학기", "여름학기", "2학기", "겨울학기" };
        semesterPicker.setMinValue(1);
        semesterPicker.setMaxValue(4);
        semesterPicker.setDisplayedValues(semesters);
        semesterPicker.setOnValueChangedListener(new NumberPicker.OnValueChangeListener() {
            @Override
            public void onValueChange(NumberPicker picker, int oldVal, int newVal) {
                semester = newVal;
            }
        });

        submitButton = (Button) findViewById(R.id.submit);
        submitButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String title = titleText.getText().toString();
                TableManager.getInstance().postTable(year, semester, title, new Callback<List<Table>>() {
                    @Override
                    public void success(List<Table> tables, Response response) {
                        // 보고있는 테이블 정보 변경하기
                        //startTableView(id);
                        finish();
                    }
                    @Override
                    public void failure(RetrofitError error) {
                        Toast.makeText(TableCreateActivity.this, "테이블 생성에 실패하였습니다.", Toast.LENGTH_SHORT).show();
                }
                });
            }
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        activityList.remove(this);
    }
}

File: app/src/main/java/com/wafflestudio/snutt2/ui/TableCreateActivity.kt
package com.wafflestudio.snutt2.ui

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.Spinner
import com.wafflestudio.snutt2.R
import com.wafflestudio.snutt2.SNUTTBaseActivity
import com.wafflestudio.snutt2.handler.ApiOnError
import com.wafflestudio.snutt2.manager.TableManager
import com.wafflestudio.snutt2.model.Coursebook
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

/**
 * Created by makesource on 2016. 3. 1..
 */
@AndroidEntryPoint
class TableCreateActivity : SNUTTBaseActivity() {
    @Inject
    lateinit var tableManager: TableManager

    @Inject
    lateinit var apiOnError: ApiOnError

    private var year = -1
    private var semester = -1
    private var titleText: EditText? = null
    private var semesterSpinner: Spinner? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        activityList.add(this)
        setContentView(R.layout.activity_table_create)
        title = "새로운 시간표"
        semesterSpinner = findViewById<View>(R.id.spinner) as Spinner
        titleText = findViewById<View>(R.id.table_title) as EditText
        tableManager.getCoursebook()
            .bindUi(
                this,
                onSuccess = { coursebooks ->
                    val displays = getDisplayList(coursebooks)
                    val years = getYearList(coursebooks)
                    val semesters = getSemesterList(coursebooks)
                    year = years[0]
                    semester = semesters[0]
                    val adapter = ArrayAdapter(baseContext, android.R.layout.simple_spinner_dropdown_item, displays)
                    semesterSpinner!!.adapter = adapter
                    semesterSpinner!!.onItemSelectedListener = object : OnItemSelectedListener {
                        override fun onItemSelected(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
                            year = years[position]
                            semester = semesters[position]
                        }

                        override fun onNothingSelected(parent: AdapterView<*>?) {}
                    }
                },
                onError = {
                    apiOnError(it)
                }
            )
    }

    override fun onDestroy() {
        super.onDestroy()
        activityList.remove(this)
    }

    private fun getDisplayList(coursebooks: List<Coursebook>): Array<String?> {
        val size = coursebooks.size
        val list = arrayOfNulls<String>(size)
        for (i in 0 until size) {
            val year: String = coursebooks[i].year.toString() + " 년"
            val semester = getSemester(coursebooks[i].semester.toInt())
            list[i] = "$year $semester"
        }
        return list
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_table_create, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.action_create) {
            val title = titleText!!.text.toString()
            // Refactoring FIXME: error handling
            tableManager.postTable(year.toLong(), semester.toLong(), title)
                .bindUi(this) {
                    finish()
                }
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun getSemester(semester: Int): String {
        return when (semester) {
            1 -> "1학기"
            2 -> "여름학기"
            3 -> "2학기"
            4 -> "겨울학기"
            else -> "(null)"
        }
    }

    private fun getYearList(coursebooks: List<Coursebook>): IntArray {
        val size = coursebooks.size
        val list = IntArray(size)
        for (i in 0 until size) {
            list[i] = coursebooks[i].year.toInt()
        }
        return list
    }

    private fun getSemesterList(coursebooks: List<Coursebook>): IntArray {
        val size = coursebooks.size
        val list = IntArray(size)
        for (i in 0 until size) {
            list[i] = coursebooks[i].semester.toInt()
        }
        return list
    }
}
